### PR TITLE
expand github feature request template

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Expands the feature-request template to include sections on API usage, and proposed implementation. This was generated using the latest version of `github-templates(1)`. Thanks!

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/feature_request.md [hunk 1]
```diff
@@ -11,9 +11,22 @@ One paragraph explanation of the feature.
...
-Tell us how the feature should work. Explain the feature largely in terms of
-examples.
+## Guide-level explanation
+Explain the proposal as if it was already included in the project and you
+were teaching it to another programmer. That generally means:
+
+- Introducing new named concepts.
+- Explaining the feature largely in terms of examples.
+- If applicable, provide sample error messages, deprecation warnings, or
+  migration guidance.
```
Comment (spacekookie) [inline .github/ISSUE_TEMPLATE/feature_request.md:21]: I would remove the `.` at the end there

File: .github/ISSUE_TEMPLATE/feature_request.md [hunk 2]
```diff
@@ -11,9 +11,22 @@ One paragraph explanation of the feature.
...
+
+## Reference-level explanation
+This is the technical portion of the feature request. Explain the design in
+sufficient detail that:
```
Comment (spacekookie) [inline .github/ISSUE_TEMPLATE/feature_request.md:25]: I think it's important to signal that feature requests are welcome, even if not all reference-level details have been resolved/ worked out.

That generally means adding some detail or section about letting users provide open or unresolved questions at this layer (not quite the same as "drawbacks" below)

File: .github/ISSUE_TEMPLATE/feature_request.md [hunk 3]
```diff
@@ -11,9 +11,22 @@ One paragraph explanation of the feature.
...
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
 
 ## Drawbacks
 Why should we _not_ do this?
```
Comment (spacekookie) [inline .github/ISSUE_TEMPLATE/feature_request.md:32]: Maybe rephrase this to "Unwanted consequences" and also make sure to mention that it's obviously just what _they_ noticed? Like, I get the feeling sometimes that people are scared of not listing all "drawbacks" and that if new ones come up, that it's essentially cancelling their feature request. Which really isn't the case most of the time.

**Review by spacekookie — COMMENTED** [review]
I was brought here by the "stale" bot, and left a review instead :P Is that how that's meant to go?

File: .github/ISSUE_TEMPLATE/feature_request.md (hunk 2 above)
Comment (yoshuawuyts) [reply .github/ISSUE_TEMPLATE/feature_request.md:25]: Yeah, I think you're right.
